src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -306,26 +308,31 @@ private void EntryExpired(CacheEntry entry)
         {
             // TODO: For efficiency consider processing these expirations in batches.
             RemoveEntry(entry);
-            StartScanForExpiredItems();
+            StartScanForExpiredItemsIfNeeded(_options.Clock.UtcNow);
         }
 
         // Called by multiple actions to see how long it's been since we last checked for expired items.
         // If sufficient time has elapsed then a scan is initiated on a background task.
-        private void StartScanForExpiredItems(DateTimeOffset? utcNow = null)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void StartScanForExpiredItemsIfNeeded(DateTimeOffset utcNow)
         {
             // Since fetching time is expensive, minimize it in the hot paths
-            DateTimeOffset now = utcNow ?? _options.Clock.UtcNow;
-            if (_options.ExpirationScanFrequency < now - _lastExpirationScan)
+            if (_options.ExpirationScanFrequency < utcNow - _lastExpirationScan)
             {
-                _lastExpirationScan = now;
+                ScheduleTask(utcNow);
+            }
+
+            void ScheduleTask(DateTimeOffset utcNow)
+            {
+                _lastExpirationScan = utcNow;
                 Task.Factory.StartNew(state => ScanForExpiredItems((MemoryCache)state), this,
                     CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
             }
         }
 
         private static void ScanForExpiredItems(MemoryCache cache)
         {
-            DateTimeOffset now = cache._options.Clock.UtcNow;
+            DateTimeOffset now = cache._lastExpirationScan = cache._options.Clock.UtcNow;
```

Comment:
Why set it both here and in ScheduleTask?

Reply:
because there can be a delay between the moment we schedule the task and the code gets actually executed

Reply:
btw @stephentoub in this particular scenario, which is more or less fire&forget, should we schedule the work item to thread pool directly?

Reply:
That's probably fine; I expect it was written this way when targeting a previous netstandard API set that didn't include ThreadPool.  You'll find two main differences:
1. If ScanForExpiredItems throws an exception, the way it's currently written that exception will get eaten.  If you change it to use ThreadPool, it'll go unhandled and crash the process.
2. The way it's currently written, if the calling thread is a thread pool thread, the queued work item will go into the local queue, whereas if you use QueueUserWorkItem, by default it'll go into the global queue.

Reply:
@stephentoub thanks for the explanation